rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll

```diff
@@ -0,0 +1,340 @@
+    exists(Pat mid |
+      mid = getAVariablePatAncestor(v) and
+      result = getImmediatePatParent(mid)
+    )
+  }
+
+  /**
+   * Holds if `v` is named `name` and is declared inside variable scope
+   * `scope`, and `v` is bound starting from `(line, column)`.
+   */
+  private predicate variableDeclInScope(
+    Variable v, VariableScope scope, string name, int line, int column
+  ) {
+    name = v.getName() and
+    exists(Pat pat | pat = getAVariablePatAncestor(v) |
+      scope =
+        any(MatchArmScope arm |
+          arm.getPat() = pat and
+          arm.getLocation().hasLocationInfo(_, line, column, _, _)
+        )
+      or
+      exists(Function f |
+        f.getParamList().getAParam().getPat() = pat and
+        scope = f.getBody() and
+        scope.getLocation().hasLocationInfo(_, line, column, _, _)
+      )
+      or
+      exists(LetStmt let |
+        let.getPat() = pat and
+        scope = getEnclosingScope(let) and
+        // for `let` statements, variables are bound _after_ the statement, i.e.
+        // not in the RHS
+        let.getLocation().hasLocationInfo(_, _, _, line, column)
+      )
+      or
+      exists(IfExpr ie, LetExpr let |
+        let.getPat() = pat and
+        ie.getCondition() = let and
+        scope = ie.getThen() and
+        scope.getLocation().hasLocationInfo(_, line, column, _, _)
+      )
+      or
+      exists(ForExpr fe |
+        fe.getPat() = pat and
+        scope = fe.getLoopBody() and
+        scope.getLocation().hasLocationInfo(_, line, column, _, _)
+      )
+      or
+      exists(ClosureExpr ce |
+        ce.getParamList().getAParam().getPat() = pat and
+        scope = ce.getBody() and
+        scope.getLocation().hasLocationInfo(_, line, column, _, _)
+      )
+    )
+  }
+
+  /**
+   * Holds if `cand` may access a variable named `name` at
+   * `(startline, startcolumn, endline, endcolumn)` in the variable scope
+   * `scope`.
+   *
+   * `nestLevel` is the number of nested scopes that need to be traversed
+   * to reach `scope` from `cand`.
+   */
+  private predicate variableAccessCandInScope(
+    VariableAccessCand cand, VariableScope scope, string name, int nestLevel, int startline,
+    int startcolumn, int endline, int endcolumn
+  ) {
+    name = cand.getName() and
+    scope = [cand.(VariableScope), getEnclosingScope(cand)] and
+    cand.getLocation().hasLocationInfo(_, startline, startcolumn, endline, endcolumn) and
+    nestLevel = 0
+    or
+    exists(VariableScope inner |
+      variableAccessCandInScope(cand, inner, name, nestLevel - 1, _, _, _, _) and
+      scope = getEnclosingScope(inner) and
+      // Use the location of the inner scope as the location of the access, instead of the
+      // actual access location. This allows us to collapse multiple accesses in inner
+      // scopes to a single entity
+      scope.getLocation().hasLocationInfo(_, startline, startcolumn, endline, endcolumn)
+    )
+  }
+
+  private newtype TVariableOrAccessCand =
```

Comment:
This type looked a little strange to me at first sight. Perhaps you could add a comment explaining the idea of algorithm. If I understand correctly you rank all variable definitions and uses, and bind each use to the highest ranking definition that has a lower rank than the variable use.

Reply:
Fair; I have tried to document the class with the algorithm that is used.